### ISSUE TITLE
Array indexes get out of order in Field Array example

### ISSUE THREAD

**Issue by thoughtassassin** [react-hook-form/react-hook-form#809]
Please consider asking the question at our spectrum channel.

https://spectrum.chat/react-hook-form

Bill, you are probably going to kill me for adding another issue. This time I marked it as question but I think it is, in fact, a bug. It looks like the field arrays get out of order. I was seeing some strange behavior in my own form and I pretty much copied what is in the example. 

**To Reproduce**
1. Go to https://codesandbox.io/s/6j1760jkjk
2. Click on 'Add Friend' twice.
3. You will see First Name: 0, First Name: 1
4. Delete First Name 0.
5. Add another friend.
6. You will see Friend 1, Friend 1.

**Codesandbox link**
https://codesandbox.io/s/6j1760jkjk

**Screenshots**
<img width="1572" alt="Screen Shot 2020-01-09 at 4 05 07 PM" src="https://user-images.githubusercontent.com/1539057/72108778-eba62080-32f9-11ea-85a4-b83a8b4aac28.png">

**Additional context**

**Comment by bluebill1049** [comment]
Lol not gonna kill u mate❤️ On the bus I will take a look today 👌

**Comment by bluebill1049** [comment]
FYI this one: https://github.com/react-hook-form/react-hook-form/pull/768

**Comment by bluebill1049** [comment]
what about this now? https://codesandbox.io/s/6j1760jkjk

**Comment by thoughtassassin** [comment]
That one seems completely broken. It solves the problem with 0 and 1 but if you add several elements and start deleting from the middle it stops working. That seems to be the problem that I am running into in the form that I working on too.

Try this on that example:

1. Add three friends. (should be 0, 1, 2)
2. Remove the second friend. (numbered 1)
3. Try to remove what is now numbered friend number 1.

At that point it no longer deletes.

**Comment by bluebill1049** [comment]
oops check this codesandbox: https://codesandbox.io/s/react-hook-form-usefieldarray-vy8fv

**Comment by bluebill1049** [comment]
or this one as well: https://codesandbox.io/s/react-hook-form-field-array-advanced-with-delete-insert-append-edit-l19pz

**Comment by thoughtassassin** [comment]
lol. I will check those out.

**Comment by thoughtassassin** [comment]
Where in the heck is useFieldArray coming from? Is that something you just wrote?

**Comment by bluebill1049** [comment]
> Where in the heck is useFieldArray coming from? Is that something you just wrote?

that's the new feature coming, still WIP :) i posted you the PR above.

**Comment by bluebill1049** [comment]
you like that new custom hook?

**Comment by thoughtassassin** [comment]
It's really slick. However, I see one issue with it. To be clear I am talking about this one: https://codesandbox.io/s/react-hook-form-usefieldarray-vy8fv.

1) Add two fields
2) Delete the top one
3) Console.log shows no values submitted even though there is one on the UI.

**Comment by thoughtassassin** [comment]
It looks like this one might be working: https://codesandbox.io/s/react-hook-form-field-array-advanced-with-delete-insert-append-edit-l19pz

I recently refactored my app to use the React Hook Version 4 Controller. I might try to see if I can refactor my field array to do what this is.

**Comment by bluebill1049** [comment]
try this latest beta: `4.5.0-beta.6` i think useFieldArray should be pretty stable

**Comment by thoughtassassin** [comment]
Thank you, once again. Is the issue that I am seeing in the comment about the customHook going to be an issue in 4.5.0~beta.6?

**Comment by bluebill1049** [comment]
> Thank you, once again. Is the issue that I am seeing in the comment about the customHook going to be an issue in 4.5.0~beta.6?

should fix it, otherwise that custom hook is rubbish lol

**Comment by thoughtassassin** [comment]
Gotta tell you, Bill. I am refactoring with the new FieldArray and I feel like I am really close. There is a lot of very complex logic in these components. Nevertheless, I was able to remove the logic for maintaining the field array and it looks like I can get everything to render. The only pesky problem is the issue with deleting indicies. 

But I'll tell you what, my friend, when you get this working I will heavily evangelize React Hook Form. The company I work for has a Frontend Guild. I'll do a talk on using React Hook Form so that other devs in my company can use this.

**Comment by bluebill1049** [comment]
> The only pesky problem is the issue with deleting indices.

Is this issue still exist in `useFieldArray`?

**Comment by thoughtassassin** [comment]
As of earlier today it does. I am using 4.5.0-beta.6. I see strange behavior when I delete. I looked at your PR for useFieldArrays and I think I was able to reproduce what simmo was talking about in the example that you shared on Codepen:

> Remove item 0
> Remove item 1
> Remove item 2
> Submit - Array empty 👍
> Append
> Append
> Append
> Submit - Array only contains 1 item, expect 3.

That is still an issue in the example you provided.

**Comment by bluebill1049** [comment]
did you change this line to `react-hook-form`? the src part.

`import { useForm, useFieldArray } from "./src";`

**Comment by bluebill1049** [comment]
nvm i see the issue, will take a look at it.

**Comment by thoughtassassin** [comment]
It's hard. You are solving a very hard problem. It was breaking my mind to keep track or those arrays. Like I said though everything else is great. I was able to cut out of a lot of code and let React Hook Form handle everything. That is exactly what I want as a developer.

**Comment by bluebill1049** [comment]
That's awesome to hear, I think that's probably the last blocker for me to resolve now :) will keep you in the loop when it's ready.

**Comment by bluebill1049** [comment]
solved: https://codesandbox.io/s/react-hook-form-usefieldarray-vy8fv

**Comment by bluebill1049** [comment]
find another edge case: back to working again...

**Comment by thoughtassassin** [comment]
That one is solved. 

I have a question about the uniqueness of names. I forked your sandbox and for the default values I replaced ```${item.name}${index}``` with```test[${index}].name``` . I thought as I randomly clicked remove on elements randomly I would start to see the sequence of arrays start over. As a result some of the names are identical. I know in HTML that the name attribute in a form doesn't have to be identical but I wonder how React Hook Form deals with that and if that isn't an issue.

To see this behavior:
1. Append two elements. You will see 0 and 1. 
2. Delete 0. 
3. Append a new element. You will see two number 1s.

This illustrates that two form elements have the same name. I'm not sure if this is an issue or if there would be ramifications for duplicate name values.

https://codesandbox.io/s/react-hook-form-usefieldarray-u8igk
